price-feeder/oracle/provider/binance.go

```diff
@@ -23,96 +26,185 @@ type (
-		return TickerPrice{}, fmt.Errorf(
-			"received unexpected symbol from Binance response; expected: %s, got: %s",
-			ticker, tickerResp.Symbol,
-		)
+	p.setTickerPair(tickerRespWS)
+}
+
+func (p *BinanceProvider) setTickerPair(ticker BinanceTicker) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.tickers[ticker.Symbol] = ticker
+}
+
+func (ticker BinanceTicker) toTickerPrice() (TickerPrice, error) {
+	return newTickerPrice("Binance", ticker.Symbol, ticker.LastPrice, ticker.Volume)
+}
+
+// subscribeTickers subscribe to all currency pairs
+func (p *BinanceProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	params := make([]string, len(cps))
+
+	for i, cp := range cps {
+		params[i] = strings.ToLower(cp.Base + cp.Quote + "@ticker")
 	}
 
-	price, err := sdk.NewDecFromStr(tickerResp.LastPrice)
-	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to parse Binance price (%s) for %s", tickerResp.LastPrice, ticker)
+	subsMsg := newBinanceSubscriptionMsg(params...)
+	return p.wsClient.WriteJSON(subsMsg)
+}
+
+func (p *BinanceProvider) handleReceivedTickers(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewMessage):
+			// time after to avoid asking for prices too frequently
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Err(err).Msg("Binance provider could not read message")
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-p.reconnectTimer.C:
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("binance provider error reconnecting")
+			}
+		}
 	}
+}
 
-	volume, err := sdk.NewDecFromStr(tickerResp.Volume)
+// reconnect closes the last WS connection and create a new one
+// A single connection to stream.binance.com is only valid for 24 hours;
+// expect to be disconnected at the 24 hour mark
+// The websocket server will send a ping frame every 3 minutes.
+// If the websocket server does not receive a pong frame back from
+// the connection within a 10 minute period, the connection will be disconnected.
+// Unsolicited pong frames are allowed.
+func (p *BinanceProvider) reconnect() error {
+	p.wsClient.Close()
+
+	p.logger.Debug().Msg("binance reconnecting websocket")
+	wsConn, _, err := websocket.DefaultDialer.Dial(p.wsURL.String(), nil)
 	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to parse Binance volume (%s) for %s", tickerResp.Volume, ticker)
+		return fmt.Errorf("error reconnect to binance websocket: %w", err)
 	}
+	p.wsClient = wsConn
 
-	return TickerPrice{Price: price, Volume: volume}, nil
+	return p.subscribeTickers(p.subscribedPairs...)
+}
+
+// newSubscriptionMsg returns a new subscription Msg
+func newBinanceSubscriptionMsg(params ...string) BinanceSubscribeMsg {
+	return BinanceSubscribeMsg{
+		Method: "SUBSCRIBE",
+		Params: params,
+		Id:     1,
```

Comment:
What meaning or significance does the ID have? It can be anything?

Reply:
It can be any number, it can be used to check if that ID message 1 had a response

Like, a response with the same ID will be sent from the provider to us

Reply:
"The id used in the JSON payloads is an unsigned INT used as an identifier to uniquely identify the messages going back and forth." Binance doc